specs/ledger/hs/src/Ledger/Update.hs

```diff
@@ -1,29 +1,248 @@
+{-# LANGUAGE TypeFamilies #-}
 
 module Ledger.Update where
 
-import Control.Lens (makeLenses)
+import Control.Lens
+import Control.State.Transition
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import GHC.Generics (Generic)
 import Numeric.Natural
+import Ledger.Core ((⨃))
+import qualified Ledger.Core as Core
 
-import Ledger.Core (SlotCount)
+import Prelude hiding (min)
 
 -- | Protocol parameters.
 --
 data PParams = PParams -- TODO: this should be a module of @cs-ledger@.
-  { _maxBkSz  :: !Natural
+  { _maxBkSz  :: Natural
   -- ^ Maximum (abstract) block size in words.
-  , _maxHdrSz :: !Natural
+  , _maxHdrSz :: Natural
   -- ^ Maximum (abstract) block header size in words.
-  , _dLiveness :: !SlotCount
+  , _maxTxSz :: Natural
+  -- ^ Maximum (abstract) transaction size in words.
+  , _maxPropSz :: Natural
+  -- ^ Maximum (abstract) update proposal size in words.
+  , _dLiveness :: Core.SlotCount
   -- ^ Delegation liveness parameter: number of slots it takes a delegation
   -- certificate to take effect.
-  , _bkSgnCntW :: !Int
+  , _bkSgnCntW :: Int
   -- ^ Size of the moving window to count signatures.
-  , _bkSgnCntT :: !Double
+  , _bkSgnCntT :: Double
   -- ^ Fraction [0, 1] of the blocks that can be signed by any given key in a
   -- window of lenght '_bkSgnCntW'. This value will be typically between 1/5
   -- and 1/4.
-  , _bkSlotsPerEpoch :: !SlotCount
+  , _bkSlotsPerEpoch :: Core.SlotCount
+  -- ^ Number of slots in an epoch
+  , _upTtl :: Core.SlotCount
+  -- ^ Update proposal TTL in slots
+  , _scriptVersion :: Natural
+  -- ^ Script version
   } deriving (Eq, Show)
 
 makeLenses ''PParams
+
+newtype UpId = UpId Int
+  deriving (Eq, Ord, Show)
+
+-- | Protocol version
+data ProtVer = ProtVer
+  { _pvMaj :: Natural
+  , _pvMin :: Natural
+  , _pvAlt :: Natural
+  } deriving (Eq, Show)
+
+makeLenses ''ProtVer
+
+newtype ApName = ApName String
+  deriving (Eq, Ord, Show)
+
+-- | Application version
+newtype ApVer = ApVer Natural
+  deriving (Eq, Num, Show)
+
+data SwVer = SwVer
+  { _svName :: ApName
+  , _svVer :: ApVer
+  } deriving (Eq, Show, Generic)
+
+makeLenses ''SwVer
+
+-- | Update proposal
+data UProp = UProp
+  { _upId :: UpId
+  , _upIssuer :: Core.VKey
+  , _upParams :: PParams
+  , _upPV :: ProtVer
+  , _upSwVer :: SwVer
+  , _upSig :: Core.Sig UProp
+  }
+
+makeLenses ''UProp
+
+inRange :: Ord a => a -> a -> a -> Bool
+inRange l u x = l <= x && x <= u
+
+-- | Test if a pair is present in a map.
+inMap :: (Ord k, Eq v) => k -> v -> Map k v -> Bool
+inMap k v m = case Map.lookup k m of
+  Just x | x == v -> True
+  _ -> False
+
+(==>) :: Bool -> Bool -> Bool
+a ==> b = not a || (a && b)
+infix 1 ==>
+
+pvCanFollow
+  :: ProtVer
+  -> ProtVer
+  -> Bool
+pvCanFollow (ProtVer mjn min an) (ProtVer mjp mip ap)
+  = (mjp, mip, ap) < (mjn, min, an)
+  && (inRange 0 1 (mjn - mjp))
+  && ((mjp == mjn) ==> (mip +1 == min))
+  && ((mjp +1 == mjn) ==> (min == 0))
```

Comment:
Nice! This looks almost like the `LaTeX` spec.  Would it make sense to define the precedence of `==>` so that we can omit the parenthesis?

Reply:
Honestly, I find precedence confusing and adds to the "stuff I have to remember when parsing in my head", so I prefer to stick with parentheses. But could do.

Reply:
Sure, that makes sense.